### PR TITLE
Feature: Posting Journal Server Core

### LINKED ISSUE #313: (Feature) Server posting journal core [Journal Vouchers]
This issue is to track progress implementing the posting journal server-side core, responsible for posting transactions to the posting journal.  The initial implementation will only involve posting `voucher`s to the posting journal, and will be expanded with utilities for posting from other tables.

**Development Steps**
- [x] Change the structure of the `voucher` table as recommended by #280 and #301.
- [x] Create the journal `core.js` file that will compute things like daily exchange rate, enterprise id, etc...
- [x] Create the journal `voucher.js` file that will perform the copy from `voucher` -> `posting_journal`.
- [x] ~~Pass the transaction from the `voucher` controller to the `journal/core`, then to `journal/voucher`, and finally attach error handling.~~
- [x] Write integration tests to ensure that the posting process works properly.
- [ ] Test edge cases - missing exchange rates, debit/credit imbalances, missing accounts, etc.

### PR DESCRIPTION
**Pull request by jniles**
This pull request introduces a core set of queries and APIs for posting data to the posting journal.  A simplified example is given for journal vouchers, though it is lacking proper handling of gain/loss on exchange.  This feature is a TODO for all journal vouchers and cash transactions.

In contrast to the bhima-1.X posting journal, this posting journal uses transaction objects to ensure data integrity and moves _all_ logic into SQL.  This ensures that we have accurate error handling, and a really clean interface for writing SQL statements.  The downside, is that each posting function (for `voucher`s, `sale`s, etc) must set a few local variables before calling the `core.setup()` method.  I think this is a worthwhile trade off, but am wiling to hear other opinions.

This PR removes all of the previous posting journal code, in favor of using a small set of utilities.  These utilities are found in `journal/core.js`.  They include both error handing and setting SQL variables.

Closes #313.

---

Hi! Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](./docs/STYLEGUIDE.md)
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub)
that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process
and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by jniles** [comment]
@sfount once this lands, someone else could build the posting procedure from `patientInvoice`.  I'll work on the `cash` posting procedure and try and sort out gain/loss on rounding.

File: client/test/e2e/patient/invoice.spec.js [hunk 1]
```diff
@@ -30,7 +30,7 @@ describe('patient invoice', function () {
     browser.get(path);
   });
 
-  it('invoices a patient for a single item', function () {
+  it.skip('invoices a patient for a single item', function () {
```
Comment (sfount) [inline client/test/e2e/patient/invoice.spec.js:33]: Are these intentionally left skipped? I assume with these should be re-implemented by the person that does the patient invoice posting core.

Comment (jniles) [inline client/test/e2e/patient/invoice.spec.js:33]: Yes, the patient invoice module relies on `journal` to post, who's code I completely removed.  This skip was introduced last minute in https://github.com/IMA-WorldHealth/bhima-2.X/pull/320/commits/d1cab30054e375073e00dc9f84e5c052a75e4d42. Whoever works on the patient invoice module will be responsible for implementing this route and unskipping the test (as well as possibly writing a few more) to prove that this whole procedure works.

File: server/controllers/finance/journal/core.js [hunk 2]
```diff
@@ -1,198 +1,156 @@
...
+ * transaction, and properly convert them into JavaScript errors to be
+ * sent to the client.
+ *
+ * @param {Error} error - the error object thrown by MySQL.
+ * @returns {Promise} handled - a rejected promise with converted error.
+ */
+exports.handler = function handler(error) {
+  let handled;
+
+  switch (error.sqlState) {
```
Comment (sfount) [inline server/controllers/finance/journal/core.js:115]: Is there a reason this uses a `switch` and not a map of codes like [this](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/config/codes.js)? This is not a problem I'm just wondering if there was a reason for the different model. Should the other be updated? 

My first thought was that this could be handled in a middleware like other database errors however having a core handler may come in useful for considering further edge cases in the future. No conclusion on this one!

**Comment by sfount** [comment]
This code looks good to me 👍  - potentially something as core as this would benefit from a wiki entry showing the flow of data from request through to database records/ error handling. I would suggest this could be done after this has been used across a number of the core modules. (Developer focussed documentation - just so when we come back to this in a few months the idea is still clear!)

Everyone should consider the use of some of the initialised variables in the core set up process as @jniles mentioned in his comments - I'm sure this will be iterated on over time. 

As I have worked on patient invoicing in the past I will take the procedure for patient invoicing, it will have to consider `subsidies` and `billing services` as well - in 2.x these are completely separate the from the invoice database record.

**Comment by jniles** [comment]
@sfount, I appreciate the review.

The wiki entry is a good idea.  I'll work on `cash` and tackle rounding while you tackle `patientInvoice`  posting with all of the considerations (`billing services` and `subsidies`), and then we'll have much more experience to write a "best practices" wiki entry.

File: server/controllers/finance/journal/core.js (hunk 2 above)
Comment (jniles) [inline server/controllers/finance/journal/core.js:115]: We can change the location of this error handler as it makes sense.  If a middleware is better, it could be easily converted to such!

I am using `switch` for it's `default` case, which allows me to do nothing if necessary.  I find that pattern less confusing than something like:

``` js
// if no entry exists for error in the map, default to old error
let handled = map[error.sqlState] || error;
```

However, if this is a pattern you prefer, I'd be happy to change it.  What do you think?
